doc/adr/0002-steppable-cek-machine.md

```diff
@@ -0,0 +1,235 @@
+# ADR 2: Steppable CEK machine
+
+Date: 2022-10
+
+## Authors
+
+Marty Stumpf <marty.stumpf@iohk.io>  
+Ziyang Liu <ziyang.liu@iohk.io>
+
+## Status
+
+Proposed
+
+## Context
+
+In order to have a minimal viable product of a debugger for Plutus, we need a CEK machine that will give us more information for debugging than our current one.
+
+Implementing the steppable CEK machine is a non-trivial task and involves some design decisions:
+
+- should the debugging machine be a separate one?
+- should we modify the existing CEK machine to support stepping, or implement something entirely new?
+- how do we ensure performance and conformance?
+
+## Decision: A separate machine
+
+We came to the conclusion that the debugging machine should be a separate one. The main reason is that we don't want to slow down the production machine.
+
+## Argument: A separate machine
+
+The debugging machine will give the most accurate information if it was the one in production. However, the drawback of that is that we have to care about its performance.
+
+We only care about performance if the steppable CEK machine will *replace* the production machine instead of being a separate one.
```

Comment:
I think there are a couple of things here:
- The "steppable machine" which is used in the debugger
- A hypothetical refactoring of the production machine which would allow its code to be used in both the debugger and the production machine.

I think it's confusing to say we were thinking of _replacing_ the production machine with the steppable one.

I think this sentence is rather: we only care about performance if the steppable machine and the production machine share code.